### PR TITLE
[FLINK-16666][python][table] Introduce Python dependency config options

### PR DESCRIPTION
**Pull request by WeiZhong94**
## What is the purpose of the change

*This pull request introduces new Python Dependency configuration options*. 

## Brief change log

  - *Introduce new Python Dependency configuration options in `PythonOptions`.*
  - *Add `PythonDependencyUtils` and use it to process the Python Dependency configuration options while the `translateToPlan()` method of Python physical node is called.*

## Verifying this change

This change is already covered by existing tests, such as *PythonOptionsTest*, *PythonDependencyUtilsTest*.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (docs)

### CODE REVIEW

File: flink-core/src/main/java/org/apache/flink/configuration/PythonOptions.java [hunk 1]
```diff
@@ -81,4 +79,60 @@
...
+		.key("python.executable")
+		.stringType()
+		.noDefaultValue()
+		.withDescription("Specify the path of the python interpreter used to execute the python " +
+			"UDF worker. The python UDF worker depends on Python 3.5+, Apache Beam " +
+			"(version == 2.19.0), Pip (version >= 7.1.0) and SetupTools (version >= 37.0.0). " +
+			"Please ensure that the specified environment meets the above requirements. The " +
+			"option is equivalent to the command line option \"-pyexec\".");
+
+	public static final ConfigOption<String> PYTHON_CLIENT_EXECUTABLE = ConfigOptions
```
Comment (HuangXingBo) [inline flink-core/src/main/java/org/apache/flink/configuration/PythonOptions.java]: Add ".stringType()" ?

File: flink-java/src/main/java/org/apache/flink/api/java/utils/PythonDependencyManager.java [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+
+		if (requirementsCachedDir != null) {
+			String cacheDirKey = generateUniqueFileKey(PYTHON_REQUIREMENTS_CACHE_PREFIX);
+			registerCachedFile(requirementsCachedDir, cacheDirKey);
+			config.get(PYTHON_REQUIREMENTS_FILE).put(CACHE, cacheDirKey);
+		}
+	}
+
+	/**
+	 * Adds a Python archive file (zip format). The file will be extracted and move to a dedicated
```
Comment (HuangXingBo) [inline flink-java/src/main/java/org/apache/flink/api/java/utils/PythonDependencyManager.java]: move -> moved

File: flink-java/src/main/java/org/apache/flink/api/java/utils/PythonDependencyManager.java [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+	 * "python.executable" could access the extracted files via relative path.
+	 *
+	 * @param archivePath The path of the archive file.
+	 */
+	public void addPythonArchive(String archivePath) {
+		addPythonArchive(archivePath, null);
+	}
+
+	/**
+	 * Adds a Python archive file (zip format). The file will be extracted and move to a dedicated
```
Comment (HuangXingBo) [inline flink-java/src/main/java/org/apache/flink/api/java/utils/PythonDependencyManager.java]: ditto

File: flink-java/src/main/java/org/apache/flink/api/java/utils/PythonDependencyManager.java [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+		}
+		if (targetDir == null) {
+			targetDir = new File(archivePath).getName();
+		}
+		if (archiveAlreadyRegistered(archivePath, targetDir)) {
+			return;
+		}
+		String fileKey = generateUniqueFileKey(PYTHON_ARCHIVE_PREFIX);
+		registerCachedFile(archivePath, fileKey);
+		if (targetDir != null) {
```
Comment (HuangXingBo) [inline flink-java/src/main/java/org/apache/flink/api/java/utils/PythonDependencyManager.java]: You have assigned targetDir, so targetDir cannot be null.

File: flink-java/src/main/java/org/apache/flink/api/java/utils/PythonDependencyManager.java [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+			}
+		}
+	}
+
+	private void registerCachedFile(String filePath, String fileKey) {
+		cachedFiles.add(new Tuple2<>(fileKey, new DistributedCache.DistributedCacheEntry(filePath, false)));
+	}
+
+	public static void synchronizePythonDependencyMetadata(Configuration from, Configuration to) {
+		from.getOptional(PYTHON_FILES).ifPresent(m -> {
```
Comment (HuangXingBo) [inline flink-java/src/main/java/org/apache/flink/api/java/utils/PythonDependencyManager.java]: Maybe we can remove the redundant {}

File: flink-java/src/main/java/org/apache/flink/api/java/utils/PythonDependencyManager.java [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+	/**
+	 * Adds a Python dependency which could be .py files, Python packages(.zip, .egg etc.) or
+	 * local directories. The dependencies will be added to the PYTHONPATH of the Python UDF worker
+	 * and the local Py4J python client.
+	 *
+	 * @param filePath The path of the Python dependency.
+	 */
+	public void addPythonFile(String filePath) {
+		Preconditions.checkNotNull(filePath);
+		if (fileAlreadyRegistered(filePath)) {
```
Comment (dianfu) [inline flink-java/src/main/java/org/apache/flink/api/java/utils/PythonDependencyManager.java]: Should we throw an exception if this case occurs as usually this is not expected?

File: flink-java/src/main/java/org/apache/flink/api/java/utils/PythonDependencyManager.java [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+	 * will be uploaded to support offline installation. These dependencies will be installed by
+	 * the command "pip install -r [requirements file] --find-links [requirements cached dir]"
+	 * before launching the Python UDF worker.
+	 *
+	 * @param requirementsFilePath The path of the requirements file.
+	 * @param requirementsCachedDir  The path of the requirements cached directory.
+	 */
+	public void setPythonRequirements(String requirementsFilePath, String requirementsCachedDir) {
+		Preconditions.checkNotNull(requirementsFilePath);
+		if (!config.contains(PYTHON_REQUIREMENTS_FILE)) {
```
Comment (dianfu) [inline flink-java/src/main/java/org/apache/flink/api/java/utils/PythonDependencyManager.java]: Ditto. Should also consider if an exception should be thrown.

File: flink-java/src/main/java/org/apache/flink/api/java/utils/PythonDependencyManager.java [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+	/**
+	 * Adds a Python archive file (zip format). The file will be extracted and move to a dedicated
+	 * directory under the working directory of the Python UDF workers. The param `targetDir` is the
+	 * name of the dedicated directory. The Python UDFs and the config option "python.executable"
+	 * could access the extracted files via relative path.
+	 *
+	 * @param archivePath The path of the archive file.
+	 * @param targetDir The name of the target directory.
+	 */
+	public void addPythonArchive(String archivePath, String targetDir) {
```
Comment (dianfu) [inline flink-java/src/main/java/org/apache/flink/api/java/utils/PythonDependencyManager.java]: Add `@Nullable` annotation for targetDir?

File: flink-java/src/main/java/org/apache/flink/api/java/utils/PythonDependencyManager.java [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+	/**
+	 * Specifies the third-party dependencies via a requirements file. The `requirementsCachedDir`
+	 * will be uploaded to support offline installation. These dependencies will be installed by
+	 * the command "pip install -r [requirements file] --find-links [requirements cached dir]"
+	 * before launching the Python UDF worker.
+	 *
+	 * @param requirementsFilePath The path of the requirements file.
+	 * @param requirementsCachedDir  The path of the requirements cached directory.
+	 */
+	public void setPythonRequirements(String requirementsFilePath, String requirementsCachedDir) {
```
Comment (dianfu) [inline flink-java/src/main/java/org/apache/flink/api/java/utils/PythonDependencyManager.java]: Add `@Nullable` annotation for requirementsCachedDir?

File: flink-java/src/main/java/org/apache/flink/api/java/utils/PythonDependencyManager.java [hunk 10]
```diff
@@ -0,0 +1,282 @@
...
+				break;
+			}
+		}
+	}
+
+	private void registerCachedFile(String filePath, String fileKey) {
+		cachedFiles.add(new Tuple2<>(fileKey, new DistributedCache.DistributedCacheEntry(filePath, false)));
+	}
+
+	public static void synchronizePythonDependencyMetadata(Configuration from, Configuration to) {
```
Comment (dianfu) [inline flink-java/src/main/java/org/apache/flink/api/java/utils/PythonDependencyManager.java]: What about add a method such as `getPythonConfiguration` and then the caller could use `from.addAll(getPythonConfiguration(to))` to copy the Python related configurations.

File: flink-core/src/main/java/org/apache/flink/configuration/PythonOptions.java [hunk 11]
```diff
@@ -81,4 +79,60 @@
...
+		.key("python.client.executable")
+		.defaultValue("python")
+		.withDescription("The python interpreter used to launch the python process when compiling " +
+			"the jobs containing Python UDFs. Equivalent to the environment variable PYFLINK_EXECUTABLE. " +
+			"The precedence is: 1. configuration in job source code. 2. environment variable. " +
```
Comment (dianfu) [inline flink-core/src/main/java/org/apache/flink/configuration/PythonOptions.java]: Do you mean the priority? If so, suggest to change precedence to priority to make it more clear:
`The priority is as following: 1. the configuration 'python.client.executable' defined in the source code; 2. the environment variable PYFLINK_EXECUTABLE; 3. the configuration 'python.client.executable' defined in flink-conf.yaml`

**Comment by WeiZhong94** [comment]
@HuangXingBo Thanks for your comments! I have updated this PR.

File: flink-java/src/main/java/org/apache/flink/api/java/utils/PythonDependencyManager.java [hunk 12]
```diff
@@ -0,0 +1,282 @@
...
+	}
+
+	/**
+	 * Adds a Python dependency which could be .py files, Python packages(.zip, .egg etc.) or
+	 * local directories. The dependencies will be added to the PYTHONPATH of the Python UDF worker
+	 * and the local Py4J python client.
+	 *
+	 * @param filePath The path of the Python dependency.
+	 */
+	public void addPythonFile(String filePath) {
```
[the remark this answers is not in the document]
Comment (WeiZhong94) [reply flink-java/src/main/java/org/apache/flink/api/java/utils/PythonDependencyManager.java]: In this PR this method is only public for test purpose. But in following PR the "add_python_file" method of Python TableEnvironment will point to this method so I think the annotation is unnecessary as it will be removed soon.

File: flink-java/src/main/java/org/apache/flink/api/java/utils/PythonDependencyManager.java (hunk 6 above)
Comment (WeiZhong94) [reply flink-java/src/main/java/org/apache/flink/api/java/utils/PythonDependencyManager.java]: This case is expected because the `configure()` method will be called when the 
`translate()` method of the Table API planners is called. As the `translate()` can be called many times in interactive programing, a python file will be registered multiple times.

**Comment by WeiZhong94** [comment]
@dianfu Thanks for your comments! I have updated this PR.

**Comment by WeiZhong94** [comment]
@flinkbot attention @aljoscha @guoweiM It would be helpful if you can have a look on this PR as it makes some changes on `flink-core`, `flink-java` and `flink-streaming-java`. :)

**Comment by WeiZhong94** [comment]
@dianfu Thanks for your comments! I have remove the unnecessary space.

**Comment by aljoscha** [comment]
I commented on the Jira issue: https://issues.apache.org/jira/browse/FLINK-16666?focusedCommentId=17070813&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17070813.

**Comment by WeiZhong94** [comment]
@dianfu @aljoscha I have updated this PR according to the discussion result on the jira issue. Please take a look.

File: flink-python/src/main/java/org/apache/flink/client/python/PythonDependencyUtils.java [hunk 13]
```diff
@@ -0,0 +1,277 @@
...
+			if (cachedFiles.stream().noneMatch(t -> t.f0.equals(fileKey))) {
+				cachedFiles.add(new Tuple2<>(fileKey, new DistributedCache.DistributedCacheEntry(filePath, false)));
+			}
+		}
+
+		private void removeCachedFileIfExists(String fileKey) {
+			cachedFiles.removeAll(cachedFiles.stream().filter(t -> t.f0.equals(fileKey)).collect(Collectors.toSet()));
+		}
+
+		public Configuration getInternalConfiguration() {
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/client/python/PythonDependencyUtils.java:273]: `Internal` is a little confusing as it doesn't explain what it contains. What about renaming it to one meaningful name, e.g. getConfigWithPythonDependencyOptions

**Comment by WeiZhong94** [comment]
@dianfu Thanks for your review! I have updated this PR. As it seems not possible to keep atomic between FLINK-16666 and FLINK-16668, I have cherry-picked the content of FLINK-16668 to this PR.

File: flink-python/src/main/java/org/apache/flink/client/python/PythonDriverEnvUtils.java [hunk 14]
```diff
@@ -256,4 +328,17 @@ public static Process startPythonProcess(PythonEnvironment pythonEnv, List<Strin
 
 		return process;
 	}
+
+	public static Process launchPy4jPythonClient(
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/client/python/PythonDriverEnvUtils.java:332]: The name is confusing as it actually starts up a Python process and the main entrypoint of the Python process is one user's Python program instead of some script provided by Py4j. So what about rename it to prepareAndStartPythonProcess or revert this abstraction if we cannot come up with a meaningful name? Besides, what about moving this method to PythonDriver?

File: flink-python/src/main/java/org/apache/flink/client/python/PythonDriver.java [hunk 15]
```diff
@@ -126,4 +119,18 @@ static GatewayServer startGatewayServer() {
...
 	}
+
+	/**
+	 * Constructs the Python configuration which will be used to prepare the environment of python process.
+	 *
+	 * @param pythonDriverOptions parsed Python command options
+	 */
+	static ReadableConfig constructPythonConfiguration(final PythonDriverOptions pythonDriverOptions) {
+		Configuration config = new Configuration();
+		config.set(
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/client/python/PythonDriver.java:130]: Could we add some description on why only PYTHON_FILES is used to construct the configuration?

File: flink-python/src/main/java/org/apache/flink/client/python/PythonDriverEnvUtils.java [hunk 16]
```diff
@@ -54,17 +64,28 @@
...
+
+	public static final String PYFLINK_CLIENT_EXECUTABLE = "PYFLINK_CLIENT_EXECUTABLE";
 
 	@VisibleForTesting
-	public static final String PYFLINK_PY_EXECUTABLE = "PYFLINK_PY_EXECUTABLE";
+	static Configuration globalConf = GlobalConfiguration.loadConfiguration();
 
 	@VisibleForTesting
-	public static final String PYFLINK_PY_ARCHIVES = "PYFLINK_PY_ARCHIVES";
+	static Map<String, String> systemEnv = System.getenv();
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/client/python/PythonDriverEnvUtils.java:77]: private?

File: flink-python/src/main/java/org/apache/flink/client/python/PythonDriverEnvUtils.java [hunk 17]
```diff
@@ -54,17 +64,28 @@
...
+
+	public static String loadContextConfiguration(
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/client/python/PythonDriverEnvUtils.java:79]: private?

File: flink-python/src/main/java/org/apache/flink/client/python/PythonDriverEnvUtils.java [hunk 18]
```diff
@@ -105,46 +126,113 @@ public void run() {
...
-			PythonDriverOptions pythonDriverOptions,
-			String tmpDir) throws IOException, InterruptedException {
+			ReadableConfig config,
+			String tmpDir) throws IOException {
+		Configuration contextConfig = ExecutionEnvironment.getExecutionEnvironment().getConfiguration();
+
 		PythonEnvironment env = new PythonEnvironment();
 
+		// 1. setup the path of python interpreter.
+		if (config.getOptional(PYTHON_CLIENT_EXECUTABLE).isPresent()) {
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/client/python/PythonDriverEnvUtils.java:141]: Revert this change as it's out of the scope of this PR?

File: flink-python/src/main/java/org/apache/flink/client/python/PythonDriverEnvUtils.java [hunk 19]
```diff
@@ -105,46 +126,113 @@ public void run() {
...
 		FileSystem fs = tmpDirPath.getFileSystem();
 		fs.mkdirs(tmpDirPath);
 
 		env.tempDirectory = tmpDir;
+
+		// 3. append the internal lib files to PYTHONPATH.
+		appendInternalLibFiles(env);
+
+		// 4. copy relevant python files to tmp dir and set them in PYTHONPATH.
+		if (contextConfig.getOptional(PythonOptions.PYTHON_FILES).isPresent()) {
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/client/python/PythonDriverEnvUtils.java:165]: Revert this change as it's out of the scope of this PR?

File: flink-python/src/main/java/org/apache/flink/client/python/PythonDriverEnvUtils.java [hunk 20]
```diff
@@ -105,46 +126,113 @@ public void run() {
...
+		}
+		if (config.getOptional(PythonOptions.PYTHON_FILES).isPresent()) {
+			List<Path> userFiles = Arrays.stream(config.get(PythonOptions.PYTHON_FILES).split(","))
+				.map(Path::new).collect(Collectors.toList());
+			PythonDriverEnvUtils.appendUserFiles(env, userFiles);
+		}
+		return env;
+	}
+
+	public static void appendInternalLibFiles(PythonEnvironment env) {
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/client/python/PythonDriverEnvUtils.java:178]: private?

File: flink-python/src/main/java/org/apache/flink/client/python/PythonDriverEnvUtils.java [hunk 21]
```diff
@@ -105,46 +126,113 @@ public void run() {
...
 		List<String> pythonPathList = new ArrayList<>();
 
-		// 2. append the internal lib files to PYTHONPATH.
-		List<java.nio.file.Path> pythonLibs = getLibFiles(FLINK_OPT_DIR_PYTHON);
+		List<java.nio.file.Path> pythonLibs;
+		if (FLINK_OPT_DIR != null && !FLINK_OPT_DIR.isEmpty()) {
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/client/python/PythonDriverEnvUtils.java:182]: revert this change?

File: flink-python/src/main/java/org/apache/flink/client/python/PythonDependencyUtils.java [hunk 22]
```diff
@@ -0,0 +1,280 @@
...
+			throw new RuntimeException(e);
+		}
+		return configurePythonDependenciesInternal(pythonDependencyManager, envConfiguration, config);
+	}
+
+	private static Configuration configurePythonDependenciesInternal(
+			PythonDependencyManager pythonDependencyManager,
+			Configuration envConfiguration,
+			Configuration config) {
+		Configuration internalConfig = pythonDependencyManager.getConfigWithPythonDependencyOptions();
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/client/python/PythonDependencyUtils.java:100]: The following code is very confusing. Is it possible to improve it?

**Comment by WeiZhong94** [comment]
@dianfu Thanks for your review! I have updated this PR according to your comments.

File: flink-python/src/test/java/org/apache/flink/client/python/PythonDependencyUtilsTest.java [hunk 23]
```diff
@@ -0,0 +1,246 @@
...
+
+	private List<Tuple2<String, DistributedCache.DistributedCacheEntry>> cachedFiles;
+
+	@Before
+	public void setUp() {
+		cachedFiles = new ArrayList<>();
+	}
+
+	@Test
+	public void testAddPythonFile() {
```
Comment (dianfu) [inline flink-python/src/test/java/org/apache/flink/client/python/PythonDependencyUtilsTest.java:54]: Rename to `testPythonFiles` as we are testing the configuration python.files instead of the method `addPythonFile`. Same for the other test cases.

File: flink-python/src/test/java/org/apache/flink/client/python/PythonDependencyUtilsTest.java [hunk 24]
```diff
@@ -0,0 +1,246 @@
...
+			"python_archive_5f3fca2a4165c7d9c94b00bfab956c15f14c41e9e03f6037c83eb61157fce09c", "py37.zip");
+		expectedConfiguration.get(PYTHON_ARCHIVES).put(
+			"python_archive_b9ceeb8434d3335ec0ba298bbd52768bf5e4d136df3e157500fe4f413a1f416c", "venv");
+		expectedConfiguration.get(PYTHON_ARCHIVES).put(
+			"python_archive_44532e54ce38086fcc235a2af12479a7305820f7f39767f0215c454df54024dd", "venv2");
+		verifyConfiguration(expectedConfiguration, actual);
+	}
+
+	@Test
+	public void testConfigure() {
```
Comment (dianfu) [inline flink-python/src/test/java/org/apache/flink/client/python/PythonDependencyUtilsTest.java:174]: It seems to me that this test case is redundant and can be removed. What do you think?

**Comment by WeiZhong94** [comment]
@dianfu Thanks for your review. I have updated according to your comments.

**Comment by dianfu** [comment]
@WeiZhong94 Thanks a lot for the update. LGTM. Will merge once the travis/azure turns green.
